Verifico que el cluster esté en estado running y ejecuto primera celda sql

In [0]:
%sql
SHOW TABLES IN samples.tpch;

database,tableName,isTemporary
tpch,customer,false
tpch,lineitem,false
tpch,nation,false
tpch,orders,false
tpch,part,false
tpch,partsupp,false
tpch,region,false
tpch,supplier,false


# Parte 1 - Primeros pasos con SELECT

El esquema TPC-H simula un negocio de ventas: clientes, pedidos, productos y países. En esta parte exploramos la tabla customer .

### Ejercicio 1.1 — Conocer el dataset

In [0]:
%sql
-- Mostrar todas las tablas disponibles en el dataset samples.tpch
SHOW TABLES IN samples.tpch;


database,tableName,isTemporary
tpch,customer,false
tpch,lineitem,false
tpch,nation,false
tpch,orders,false
tpch,part,false
tpch,partsupp,false
tpch,region,false
tpch,supplier,false


In [0]:
%sql
-- Ver estructura de columnas y tipos de datos de la tabla customer
DESCRIBE samples.tpch.customer;


col_name,data_type,comment
c_custkey,bigint,null
c_name,string,null
c_address,string,null
c_nationkey,bigint,null
c_phone,string,null
c_acctbal,"decimal(18,2)",null
c_mktsegment,string,null
c_comment,string,null


#### Mostrar los 10 primeros registros

In [0]:
%sql
SELECT * 
FROM samples.tpch.customer
LIMIT 10;


c_custkey,c_name,c_address,c_nationkey,c_phone,c_acctbal,c_mktsegment,c_comment
412445,Customer#000412445,"0QAB3OjYnbP6mA0B,kgf",21,31-421-403-4333,5358.33,BUILDING,arefully blithely regular epi
412446,Customer#000412446,"5u8MSbyiC7J,7PuY4Ivaq1JRbTCMKeNVqg",20,30-487-949-7942,9441.59,MACHINERY,"sleep according to the fluffily even forges. fluffily careful packages after the ironic, silent deposi"
412447,Customer#000412447,HC4ZT62gKPgrjr ceoaZgFOunlUogr7GO,7,17-797-466-6308,7868.75,AUTOMOBILE,aggle blithely among the carefully express excus
412448,Customer#000412448,hJok1MMrDgH,6,16-541-510-4964,6060.98,MACHINERY,ly silent requests boost slyly. express courts sleep according to the fluf
412449,Customer#000412449,"zAt1nZNG01gOhIqgyDtDa S,Y0VSofZJs1dd",14,24-710-983-5536,4973.84,HOUSEHOLD,"refully final theodolites. final, slow excuses sleep quickly! quickly ironic idea"
412450,Customer#000412450,fUD6IoGdtF,20,30-293-696-5047,4406.28,BUILDING,refully final dolphins after the carefully bold packages sleep quickly express deposits. fluffily
412451,Customer#000412451,W2Ge0Qd8adH,20,30-590-724-6711,2290.38,BUILDING,slow asymptotes will are carefully final packages. slyly regular fox
412452,Customer#000412452,Ij4xiPIeNEP1uR5p7H,10,20-492-590-3363,3426.64,AUTOMOBILE,sleep slyly after the sometimes even ideas. slyly express theodolites dazzle furiously ironic dependenci
412453,Customer#000412453,4DmSxDPMmfidKQB3W50FIzkjZESEW3LPgLBuQbic,21,31-480-724-9665,4592.14,MACHINERY,"against the slyly regular requests-- pending, pending accounts boost quic"
412454,Customer#000412454,ZQfKDMUyEfn,9,19-898-261-2669,2035.91,FURNITURE,quickly. blithely special theodolites about the excus


### Ejercicio 1.2 — Seleccionar columnas específicas

In [0]:
%sql
SELECT 
  c_name AS nombre_cliente,
  c_nationkey AS codigo_pais,
  c_acctbal AS saldo_cuenta
FROM samples.tpch.customer;


nombre_cliente,codigo_pais,saldo_cuenta
Customer#000412445,21,5358.33
Customer#000412446,20,9441.59
Customer#000412447,7,7868.75
Customer#000412448,6,6060.98
Customer#000412449,14,4973.84
Customer#000412450,20,4406.28
Customer#000412451,20,2290.38
Customer#000412452,10,3426.64
Customer#000412453,21,4592.14
Customer#000412454,9,2035.91


### Ejercicio 1.3 — Contar y resumir

#### Calculamos cantidad total de clientes, saldo promedio, máximo y mínimo.

In [0]:
%sql
-- Calcular cantidad total de clientes y resumen de saldos
SELECT 
  COUNT(*) AS total_clientes,
  AVG(c_acctbal) AS saldo_promedio,
  MAX(c_acctbal) AS saldo_maximo,
  MIN(c_acctbal) AS saldo_minimo
FROM samples.tpch.customer;


total_clientes,saldo_promedio,saldo_maximo,saldo_minimo
750000,4501.027983,9999.99,-999.99


### Ejercicio 1.4 — Columnas calculadas

##### Creamos una columna `tipo_cliente` que clasifica según el saldo: 'VIP' si supera 5000, 'estándar' si no.

In [0]:
%sql
-- Clasificar clientes según su saldo
SELECT 
  c_name AS nombre_cliente,
  c_acctbal AS saldo_cuenta,
  CASE 
    WHEN c_acctbal > 5000 THEN 'VIP'
    ELSE 'estándar'
  END AS tipo_cliente
FROM samples.tpch.customer
ORDER BY saldo_cuenta DESC;


nombre_cliente,saldo_cuenta,tipo_cliente
Customer#000508503,9999.99,VIP
Customer#000061453,9999.99,VIP
Customer#000399453,9999.97,VIP
Customer#000069321,9999.96,VIP
Customer#000348586,9999.96,VIP
Customer#000242308,9999.96,VIP
Customer#000652672,9999.93,VIP
Customer#000432876,9999.91,VIP
Customer#000171663,9999.90,VIP
Customer#000577788,9999.87,VIP


##### Como solo vemos "VIPs", verificamos si trae los datos de salario maximos y minimos

In [0]:
%sql
SELECT 
  MIN(c_acctbal) AS saldo_minimo,
  MAX(c_acctbal) AS saldo_maximo
FROM samples.tpch.customer;


saldo_minimo,saldo_maximo
-999.99,9999.99


##### Buscamos los clientes "estandar"

In [0]:
%sql
SELECT 
  c_name AS nombre_cliente,
  c_acctbal AS saldo_cuenta,
  CASE 
    WHEN c_acctbal > 5000 THEN 'VIP'
    ELSE 'estándar'
  END AS tipo_cliente
FROM samples.tpch.customer
ORDER BY saldo_cuenta ASC;


nombre_cliente,saldo_cuenta,tipo_cliente
Customer#000148887,-999.99,estándar
Customer#000054020,-999.98,estándar
Customer#000007011,-999.95,estándar
Customer#000263826,-999.95,estándar
Customer#000168833,-999.95,estándar
Customer#000323237,-999.94,estándar
Customer#000123560,-999.93,estándar
Customer#000622213,-999.92,estándar
Customer#000570165,-999.90,estándar
Customer#000724804,-999.90,estándar


> Nota: El rango de saldos va de -999.99 a 9999.99, por lo que existen clientes con saldo negativo (deudores). 
> Al ordenar por saldo descendente, solo se ven los clientes con saldo alto (VIP). 
> Para visualizar los 'estándar', se puede ordenar ascendentemente o ajustar el umbral.


### Conclusion Parte 1

En esta primera parte exploramos el dataset `samples.tpch`, comprendiendo su estructura y aplicando consultas básicas.
Aprendimos a:
- Listar tablas y esquemas.
- Seleccionar columnas con alias.
- Calcular estadísticas agregadas.
- Crear columnas calculadas con `CASE WHEN`.
Estas operaciones son la base para análisis más complejos en SQL.

# Parte 2 - Filtros y agrupaciones 

La tabla orders registra cada pedido. La tabla lineitem detalla cada línea de producto
dentro de un pedido.

![image_1780267883797.png](./image_1780267883797.png "image_1780267883797.png")

En esta sección aplicamos filtros, agrupaciones y condiciones sobre el dataset `samples.tpch` para analizar pedidos y productos.

### Ejercicio 2.1 — Filtrar pedidos


In [0]:
%sql
-- Mostrar pedidos con estado 'F' y monto total mayor a 10000
SELECT 
  o_orderkey AS clave_pedido,
  o_custkey AS cliente,
  o_orderstatus AS estado,
  o_totalprice AS monto_total
FROM samples.tpch.orders
WHERE o_orderstatus = 'F' AND o_totalprice > 10000
ORDER BY o_totalprice DESC;


clave_pedido,cliente,estado,monto_total
18869634,565334,F,565520.21
4806726,183092,F,537776.40
23245603,106226,F,531353.33
19806371,439738,F,525060.79
735366,61255,F,523028.09
4267751,537949,F,520697.30
3187749,306940,F,518502.85
20567877,122161,F,518390.20
19094241,144877,F,517539.03
29167494,105110,F,516515.21


### Ejercicio 2.2 — Agrupar pedidos por estado

Contamos cuántos pedidos hay por cada estado y sumamos el monto total de cada grupo.

In [0]:
%sql
SELECT 
  o_orderstatus AS estado,
  COUNT(*) AS cantidad_pedidos,
  SUM(o_totalprice) AS monto_total
FROM samples.tpch.orders
GROUP BY o_orderstatus;


estado,cantidad_pedidos,monto_total
P,191189,35333603011.43
F,3654501,549181919365.27
O,3654310,548923692869.55


### Ejercicio 2.3 — Filtrar grupos con HAVING

Mostramos solo los estados con más de 10000 pedidos.

In [0]:
%sql
SELECT 
  o_orderstatus AS estado,
  COUNT(*) AS cantidad_pedidos,
  SUM(o_totalprice) AS monto_total
FROM samples.tpch.orders
GROUP BY o_orderstatus
HAVING COUNT(*) > 10000;


estado,cantidad_pedidos,monto_total
P,191189,35333603011.43
F,3654501,549181919365.27
O,3654310,548923692869.55


**Diferencia entre WHERE y HAVING:**  
`WHERE` filtra filas **antes** de agrupar los datos, mientras que `HAVING` filtra **después** de aplicar la agrupación.


### Ejercicio 2.4 — Top 10 productos más pedidos

Usamos la tabla `lineitem` para encontrar los 10 productos con mayor cantidad total pedida.

In [0]:
%sql
SELECT 
  l_partkey AS producto,
  SUM(l_quantity) AS cantidad_total
FROM samples.tpch.lineitem
GROUP BY l_partkey
ORDER BY cantidad_total DESC
LIMIT 10;


producto,cantidad_total
6883,1659.00
104079,1628.00
911156,1600.00
735075,1582.00
228489,1558.00
724502,1555.00
367234,1554.00
829681,1551.00
640765,1545.00
737695,1541.00


#### Conclusiones - Parte 2

En esta parte aplicamos filtros y agrupaciones para analizar los pedidos y productos del dataset `samples.tpch`.  
Aprendimos a:
- Usar `WHERE` y `AND` para filtrar filas.
- Agrupar con `GROUP BY` y resumir con `COUNT` y `SUM`.
- Aplicar condiciones sobre grupos con `HAVING`.
- Ordenar resultados y limitar la cantidad de filas con `ORDER BY` y `LIMIT`.
Estas consultas permiten obtener información resumida y segmentada de grandes volúmenes de datos.

# Parte 3 - Combinando tablas con JOIN

Los clientes tienen c_nationkey que se relaciona con n_nationkey en nation . Los pedidos
se vinculan por o_custkey = c_custkey .


En esta sección unimos tablas del dataset `samples.tpch` para relacionar clientes, pedidos y países, aplicando `JOIN`, `GROUP BY` y funciones de agregación.

### Ejercicio 3.1 — Primer JOIN: clientes y sus pedidos

Unimos `customer` con `orders` para mostrar el nombre del cliente y el monto de cada pedido.

In [0]:
%sql
-- Unir clientes con pedidos para ver nombre y monto
SELECT 
  c.c_name AS nombre_cliente,
  o.o_orderkey AS clave_pedido,
  o.o_totalprice AS monto_total
FROM samples.tpch.customer c
INNER JOIN samples.tpch.orders o
  ON c.c_custkey = o.o_custkey
LIMIT 20;


nombre_cliente,clave_pedido,monto_total
Customer#000179329,11396166,187683.10
Customer#000473245,11396167,117554.58
Customer#000077549,11396192,47611.56
Customer#000610651,11396193,121869.26
Customer#000193169,11396194,241066.47
Customer#000580097,11396195,59578.69
Customer#000432472,11396196,186262.50
Customer#000287354,11396197,182596.64
Customer#000299479,11396198,98992.78
Customer#000091625,11396199,336890.16


### Ejercicio 3.2 — Clientes por país

Unimos `customer` con `nation` para mostrar el nombre del cliente y el país.

In [0]:
%sql
-- Unir clientes con países usando la clave foránea
SELECT 
  c.c_name AS nombre_cliente,
  n.n_name AS pais
FROM samples.tpch.customer c
INNER JOIN samples.tpch.nation n
  ON c.c_nationkey = n.n_nationkey;


nombre_cliente,pais
Customer#000412445,VIETNAM
Customer#000412446,SAUDI ARABIA
Customer#000412447,GERMANY
Customer#000412448,FRANCE
Customer#000412449,KENYA
Customer#000412450,SAUDI ARABIA
Customer#000412451,SAUDI ARABIA
Customer#000412452,IRAN
Customer#000412453,VIETNAM
Customer#000412454,INDONESIA


### Ejercicio 3.3 — Total gastado por país

Combinamos `customer`, `orders` y `nation` para calcular cuánto gastaron en total los clientes de cada país.

In [0]:
%sql
-- Calcular gasto total por país
SELECT 
  n.n_name AS pais,
  SUM(o.o_totalprice) AS gasto_total
FROM samples.tpch.customer c
INNER JOIN samples.tpch.orders o
  ON c.c_custkey = o.o_custkey
INNER JOIN samples.tpch.nation n
  ON c.c_nationkey = n.n_nationkey
GROUP BY n.n_name
ORDER BY gasto_total DESC;


pais,gasto_total
IRAQ,45968155784.04
INDONESIA,45921526909.21
GERMANY,45725194459.72
FRANCE,45713656960.12
IRAN,45708446850.96
UNITED KINGDOM,45677337591.71
ETHIOPIA,45602005018.40
JORDAN,45578873952.07
ARGENTINA,45540967378.41
RUSSIA,45453202329.82


### Ejercicio 3.4 — Desafío final


Encontramos el cliente que más gastó en total, mostrando su nombre, país y monto.


In [0]:
%sql
-- Cliente que más gastó en total
SELECT 
  c.c_name AS nombre_cliente,
  n.n_name AS pais,
  SUM(o.o_totalprice) AS gasto_total
FROM samples.tpch.customer c
INNER JOIN samples.tpch.orders o
  ON c.c_custkey = o.o_custkey
INNER JOIN samples.tpch.nation n
  ON c.c_nationkey = n.n_nationkey
GROUP BY c.c_name, n.n_name
ORDER BY gasto_total DESC
LIMIT 1;


nombre_cliente,pais,gasto_total
Customer#000721180,ROMANIA,7147854.45


# Parte Bonus

Usamos el dataset `samples.nyctaxi.trips` para contar la cantidad de viajes por año en Nueva York.
Esto permite observar la evolución del volumen de viajes a lo largo del tiempo.

In [0]:
%sql
-- ¿Cuántos viajes hay por año?
SELECT year(tpep_pickup_datetime) AS anio, COUNT(*) AS viajes
FROM samples.nyctaxi.trips
GROUP BY year(tpep_pickup_datetime)
ORDER BY anio;


anio,viajes
2016,21932


> En este análisis se observa cómo varía la cantidad de viajes por año, lo que puede reflejar cambios en la demanda de transporte o en la disponibilidad de datos del dataset.


# Conclusión Final


> En conclusión, tanto el país líder (Irak) en gasto como el cliente con mayor monto total (Customer#000721180) son relevantes porque concentran recursos y permiten identificar patrones de consumo que no se observan en otros casos.


In [0]:
%sql
-- País que mas gastó
SELECT 
  n.n_name AS pais,
  SUM(o.o_totalprice) AS gasto_total
FROM samples.tpch.customer c
INNER JOIN samples.tpch.orders o
  ON c.c_custkey = o.o_custkey
INNER JOIN samples.tpch.nation n
  ON c.c_nationkey = n.n_nationkey
GROUP BY n.n_name
ORDER BY gasto_total DESC
LIMIT 1;


pais,gasto_total
IRAQ,45968155784.04


In [0]:
%sql
-- Cliente que más gastó
SELECT 
  c.c_name AS nombre_cliente,
  n.n_name AS pais,
  SUM(o.o_totalprice) AS gasto_total
FROM samples.tpch.customer c
INNER JOIN samples.tpch.orders o
  ON c.c_custkey = o.o_custkey
INNER JOIN samples.tpch.nation n
  ON c.c_nationkey = n.n_nationkey
GROUP BY c.c_name, n.n_name
ORDER BY gasto_total DESC
LIMIT 1;


nombre_cliente,pais,gasto_total
Customer#000721180,ROMANIA,7147854.45
